src/query/functions/linear/histogram_quantile.go

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package linear
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+	"github.com/m3db/m3/src/query/ts"
+)
+
+const (
+	// HistogramQuantileType calculates the quantile for histogram buckets.
+	//
+	// NB: each sample must contain a tag with a bucket name (given by tag
+	// options) that denotes the upper bound of that bucket; series without this
+	// tag are ignored.
+	HistogramQuantileType = "histogram_quantile"
+	initIndexBucketLength = 10
+)
+
+// NewHistogramQuantileOp creates a new histogram quantile operation.
+func NewHistogramQuantileOp(
+	args []interface{},
+	opType string,
+) (parser.Params, error) {
+	if len(args) != 1 {
+		return emptyOp, fmt.Errorf(
+			"invalid number of args for histogram_quantile: %d", len(args))
+	}
+
+	if opType != HistogramQuantileType {
```

Comment:
Check this first

Reply:
Was copying the pattern [from clamp.go](https://github.com/m3db/m3/blob/master/src/query/functions/linear/clamp.go#L49)

Wouldn't mind dealing with these correctly within the parser in a later pr, like the binary and aggregation functions are done at the moment.